lib/src/Pos/Util/UnitsOfMeasure.hs

```diff
@@ -22,6 +38,79 @@ data UnitOfMeasure =
     | Bytes
     | Lovelace
     | LovelacePerByte
-    deriving (Show, Eq)
+    deriving (Show, Eq, Ord)
+
+instance Buildable UnitOfMeasure where
+    build = \case
+        Bytes           -> "bytes"
+        LovelacePerByte -> "lovelace/byte"
+        Lovelace        -> "lovelace"
+        Seconds         -> "seconds"
+        Milliseconds    -> "milliseconds"
+        Microseconds    -> "microseconds"
+        Percentage100   -> "percent"
+        Blocks          -> "blocks"
+        BlocksPerSecond -> "blocks/second"
+
+instance ToJSON UnitOfMeasure where
+    toJSON = String . T.toStrict . B.toLazyText . build
+
+-- | Represent data with a given unit of measure
+data MeasuredIn (u :: UnitOfMeasure) a
+    = MeasuredIn a
+    deriving (Show, Eq, Ord)
+
+instance (Demote u, Buildable a) => BuildableSafeGen (MeasuredIn u a) where
+    buildSafeGen _ = build
+
+instance (Demote u, Buildable a) => Buildable (MeasuredIn u a) where
+    build (MeasuredIn a) = F.bprint
+        (F.build % " " % F.build)
+        a
+        (demote $ Proxy @u)
+
+instance (Demote u, ToJSON a) => ToJSON (MeasuredIn u a) where
+    toJSON (MeasuredIn a) = object
+        [ "unit"     .= demote (Proxy @u)
+        , "quantity" .= toJSON a
+        ]
+
+instance (Demote u, FromJSON a) => FromJSON (MeasuredIn u a) where
+    parseJSON = withObject "MeasuredIn" $ \o -> do
+        verifyUnit =<< o .: "unit"
+        MeasuredIn <$> o .: "quantity"
+      where
+        verifyUnit = \case
+            u@(String _) | u == toJSON (demote $ Proxy @u) -> pure ()
+            _ -> mempty
+
+instance (Demote u, ToSchema a) => ToSchema (MeasuredIn u a) where
+    declareNamedSchema _ = do
+        NamedSchema _ schema <- declareNamedSchema (Proxy @a)
+        pure $ NamedSchema (Just "MeasuredIn") $ mempty
+            & type_ .~ SwaggerObject
+            & required .~ ["quantity", "unit"]
+            & properties .~ (mempty
+                & at "quantity" ?~ Inline schema
+                & at "unit" ?~ (Inline $ mempty
+                    & type_ .~ SwaggerString
+                    & enum_ ?~ [toJSON $ demote $ Proxy @u]
+                    )
+                )
+
+--
+-- Internal
+--
 
-data MeasuredIn (a :: UnitOfMeasure) b = MeasuredIn b deriving (Eq, Show)
+-- | Bring a type back to the world of value (invert of promote)
+class Demote (u :: UnitOfMeasure) where
+    demote :: Proxy u -> UnitOfMeasure
+instance Demote 'Bytes           where demote _ = Bytes
+instance Demote 'LovelacePerByte where demote _ = LovelacePerByte
+instance Demote 'Lovelace        where demote _ = Lovelace
+instance Demote 'Seconds         where demote _ = Seconds
+instance Demote 'Milliseconds    where demote _ = Milliseconds
+instance Demote 'Microseconds    where demote _ = Microseconds
+instance Demote 'Percentage100   where demote _ = Percentage100
+instance Demote 'Blocks          where demote _ = Blocks
+instance Demote 'BlocksPerSecond where demote _ = BlocksPerSecond
```

Comment:
It's really cool to see how you're doing this!

I partly tried going down this lane previously, but was too unsure about the all the details and direction. Think I did a less general `UnitToText u` and later thought I still would have needed a "singletons" package, instead of just doing this!

Reply:
It's a pity that Haskell don't have a native / built-in to demote types yet, I believe there are good reasons for this (I never actually searched that seriously).